### ISSUE TITLE
Type does not satisfy the constraint '"__typename"'

### ISSUE THREAD

**Issue by janhartmann** [dotansimha/graphql-code-generator#1762]
**Describe the bug**
Using the compability plugin together with fragments is generating following error:

```tsx
export type CanErrorEventInlineFragment = {
    __typename: "CanErrorEvent";
  } & Pick<
    GetMachineActiveEventsQuery["machine"]["unit"]["events"][0]["details"],
    "suspectParameterNumber" // ERROR: Type '"suspectParameterNumber"' does not satisfy the constraint '"__typename"'.ts(2344)
  >;
```

The generated `GetMachineActiveEventsQuery` is:

```tsx
export type GetMachineActiveEventsQuery = { __typename?: "Query" } & {
  machine: Maybe<
    { __typename?: "Machine" } & Pick<Machine, "id"> & {
        unit: Maybe<
          { __typename?: "Unit" } & Pick<Unit, "id"> & {
              events: Maybe<
                Array<
                  { __typename?: "UnitEvent" } & Pick<
                    UnitEvent,
                    | "id"
                    | "type"
                    | "timeOn"
                    | "criticality"
                    | "description"
                    | "severity"
                  > & {
                      details: Maybe<

                          | ({ __typename?: "CanErrorEvent" } & Pick<
                              CanErrorEvent,
                              "suspectParameterNumber"
                            >)
                          | ({ __typename?: "DamageReportEvent" } & Pick<
                              DamageReportEvent,
                              "state"
                            >)
                          | ({ __typename?: "PrecheckEvent" } & Pick<
                              PrecheckEvent,
                              "count" | "failedCount"
                            >)
                          | ({ __typename?: "ServiceKmEvent" } & Pick<
                              ServiceKmEvent,
                              "serviceKm" | "currentKm"
                            >)
                          | ({ __typename?: "ServiceHourEvent" } & Pick<
                              ServiceHourEvent,
                              "serviceRun" | "currentRun"
                            >)
                          | ({ __typename?: "ServiceCalendarEvent" } & Pick<
                              ServiceCalendarEvent,
                              "date"
                            >)
                      >;
                    }
                >
              >;
            }
        >;
      }
  >;
};
```

**To Reproduce**

1. My GraphQL schema:

```graphql
schema {
  query: Query
}

type Query {
  machine: Machine
}

type Machine {
  id: Int!
  unit: Unit
}

type Unit {
  id: Int!
  events(active: Boolean!): [UnitEvent!]
}

enum UnitEventType {
  DAMAGE_REPORT
  PRE_CHECK
  SERVICE_KM
  SERVICE_HOUR
  SERVICE_HOUR2
  SERVICE_CALENDAR
  CAN_ERROR
  UNKNOWN
}

type UnitEvent {
  id: String!
  details: UnitEventDetails
}

union UnitEventDetails =
    DamageReportEvent
  | CanErrorEvent
  | PrecheckEvent
  | ServiceCalendarEvent
  | ServiceHourEvent
  | ServiceKmEvent

type DamageReportEvent {
  state: Int
}

type CanErrorEvent {
  id: String
  canErrorId: String
  suspectParameterNumber: Int
  failureModeIdentifier: Int
  canbusNumber: Int
  sourceAddress: Int
  unitId: Int
}

type PrecheckEvent {
  count: Int
  failedCount: Int
}

type ServiceCalendarEvent {
  date: String
}

type ServiceHourEvent {
  serviceRun: Int
  currentRun: Int
}

type ServiceKmEvent {
  serviceKm: Int
  currentKm: Int
}
```

2. My GraphQL operations:

```graphql
query GetMachineActiveEvents($id: Int!) {
  machine(input: { id: $id }) {
    id
    unit {
      id
      events(active: true) {
        id
        details {
          ... on CanErrorEvent {
            suspectParameterNumber
          }
          ... on DamageReportEvent {
            state
          }
          ... on PrecheckEvent {
            count
            failedCount
          }
          ... on ServiceKmEvent {
            serviceKm
            currentKm
          }
          ... on ServiceHourEvent {
            serviceRun
            currentRun
          }
          ... on ServiceCalendarEvent {
            date
          }
        }
      }
    }
  }

```

3. My `codegen.yml` config file:

```yml
schema: "server/src/schema/schema.graphql"
documents:
  - "client/src/**/*.graphql"
overwrite: true
generates:
  client/src/data/generated/types.tsx:
    config:
      scalars:
        Date: Date
        DateTime: Date
    plugins:
      - typescript
      - typescript-operations
      - typescript-react-apollo
      - typescript-compatibility
      - fragment-matcher
```

**Expected behavior**
I expected the plugin to not throw and error

**Environment:**

- OS: MacOS
```
"@graphql-codegen/cli": "^1.1.1",
"@graphql-codegen/fragment-matcher": "^1.1.1",
"@graphql-codegen/typescript-compatibility": "^1.1.1",
"@graphql-codegen/typescript-operations": "^1.1.1",
"@graphql-codegen/typescript-react-apollo": "^1.1.1",
"@graphql-codegen/typescript-resolvers": "^1.1.1",
"graphql": "^14.2.1",
```
- NodeJS: v11.14.0

**Additional context**

**Comment by dotansimha** [comment]
Hi @janhartmann , sorry for the delay, it was a difficult one :) 
The issue here is the fact that `suspectParameterNumber` might not exists, because it's existence depends on the `__typename=CanErrorEvent` (because we generate ` | ` between inline fragments of different types). 
So we have to force the `__typename=CanErrorEvent` and this way to extract only the relevant fields from the complex type. 
I added `DiscriminateUnion` helper that should cover that in case on GraphQL `union`s, by extracting the actual type using the type condition of `__typename`.

I added the fix to this PR: https://github.com/dotansimha/graphql-code-generator/pull/1798 

Can you please try the latest alpha? It should be `1.1.2-alpha-6668a794.20`.

**Comment by dotansimha** [comment]
Fixed in 1.1.3 🎉

**Comment by janhartmann** [comment]
Works! :-)
